crates/adapters/src/controller.rs

```diff
@@ -1432,97 +1422,148 @@ struct CircuitThread {
             step,
             input_metadata,
             input_statistics,
             output_statistics,
-        } = ControllerInit::new(config.clone(), storage)?;
+            pipeline_diff,
+        } = controller_init;
+
         let storage = circuit_config
             .storage
             .as_ref()
-            .map(|storage| storage.backend.clone());
+            .map(|storage: &CircuitStorageConfig| storage.backend.clone());
 
         let (mut circuit, catalog) = circuit_factory(circuit_config)?;
 
-        // Determine whether the pipeline has been modified since the checkpoint was taken.
-        //
-        // The pipeline is considered modified if:
-        // 1. The circuit has changed, triggering bootstrapping.
-        // 2. The set of output connectors has changed (modulo transient connectors like HTTP and ad hoc)
-        //
-        // Note that this does not include changes to the circuit that don't trigger bootstrapping. This is
-        // the case for changes that only remove parts of the circuit.
-        //
-        // If the pipeline has been modified, we must pick up the new connector config and disable replay,
-        // since we can't replay into a modified pipeline.
-
-        let old_configured_inputs = pipeline_config
-            .inputs
-            .iter()
-            .filter(|(_, cfg)| !cfg.connector_config.transport.is_transient())
-            .map(|(name, cfg)| (name.clone(), cfg.clone()))
-            .collect::<BTreeMap<_, _>>();
-
-        let old_configured_outputs = pipeline_config
-            .outputs
-            .iter()
-            .filter(|(_, cfg)| !cfg.connector_config.transport.is_transient())
-            .map(|(name, cfg)| (name.clone(), cfg.clone()))
-            .collect::<BTreeMap<_, _>>();
-
-        let inputs_modified = !maps_equal_with(&old_configured_inputs, &config.inputs, |a, b| {
-            a.connector_config.equal_modulo_paused(&b.connector_config)
-        });
+        let lir = circuit.lir()?;
 
-        let outputs_modified =
-            !maps_equal_with(&old_configured_outputs, &config.outputs, |a, b| {
-                a.connector_config.equal_modulo_paused(&b.connector_config)
-            });
+        if let Some(state) = &state {
+            if let Some(diff) = &pipeline_diff {
+                if let Some(bootstrap_info) = circuit.bootstrap_info() {
+                    let non_materialized_tables =
+                        non_materialized_replay_sources(bootstrap_info, &pipeline_config, diff);
+                    if !non_materialized_tables.is_empty() {
+                        return Err(ControllerError::BootstrapNotAllowed {
+                            error: format!(
+                                "- The following tables are not materialized, and at least one of the views that depend on each of these tables has changed and requires bootstrapping: {}. We recommend materializing all tables in the program to avoid such errors",
```

Comment:
this recommendation is coming a bit late; the user cannot really act on it.
Maybe "in the future?"